### PR TITLE
Issue #3966: Dashboard module conflicts with Block Reference Module

### LINKED ISSUE #3966: Dashboard module conflicts with Block Reference Module
When attempting to view a node type containing a block reference field (with a value), I am getting the following error:

"Error: Call to a member function hasContexts() on null in dashboard_preprocess_block() (line 218..."

dashboard_preprocess_block function in dashboard.module @line 217 to 218 is:

```
$layout = $variables['layout'];
if ($layout->hasContexts(array('dashboard'))) {
```
The problem with this appears to be that a block called through a block ref field may not have a layout assigned.  This will result in $layout being Null, which causes the error.

Having a null check at the beginning of the if statement @line 218 solves this issue:

```
if (!is_null($layout) && $layout->hasContexts(array('dashboard'))) {
```

---

PR by @olafgrabienski: https://github.com/backdrop/backdrop/pull/2839

### PR DESCRIPTION
**Pull request by olafgrabienski**
Fixes https://github.com/backdrop/backdrop-issues/issues/3966.

### CODE REVIEW

**Review by herbdool — CHANGES_REQUESTED** [review]
Can you make it `is_object($layout)` so it checks for a layout object. Or even better, make it `is_a($layout, 'Layout')` to ensure it is the right class.

**Comment by olafgrabienski** [comment]
> Can you (...) make it `is_a($layout, 'Layout')` to ensure it is the right class

@herbdool I've not found `is_a` in the API documentation. Are you referring to `isA` ([LayoutContext::isA](https://api.backdropcms.org/api/backdrop/core%21modules%21layout%21plugins%21context%21layout_context.inc/function/LayoutContext%3A%3AisA/1))?

I'm also not sure where exactly to put your suggestion. If you have maintainer edit permissions, feel free to edit the PR directly.

File: core/modules/dashboard/dashboard.module [hunk 1]
```diff
@@ -222,7 +222,7 @@ function dashboard_preprocess_layout(&$variables) {
 function dashboard_preprocess_block(&$variables) {
   /* @var Layout $layout */
   $layout = $variables['layout'];
-  if ($layout->hasContexts(array('dashboard'))) {
+  if (!is_null($layout) && $layout->hasContexts(array('dashboard'))) {
```
Comment (herbdool) [inline core/modules/dashboard/dashboard.module]:
```suggestion
  if (is_a($layout, 'Layout') && $layout->hasContexts(array('dashboard'))) {
```

Comment (herbdool) [reply core/modules/dashboard/dashboard.module]: @olafgrabienski I've added the suggested change. It is a core PHP function https://www.php.net/manual/en/function.is-a.php

**Comment by olafgrabienski** [comment]
@herbdool Thanks for the information. I've 'committed' your suggestion.

**Review by herbdool — APPROVED** [review]

File: core/modules/dashboard/dashboard.module [hunk 2]
```diff
@@ -222,7 +222,7 @@ function dashboard_preprocess_layout(&$variables) {
 function dashboard_preprocess_block(&$variables) {
   /* @var Layout $layout */
   $layout = $variables['layout'];
-  if ($layout->hasContexts(array('dashboard'))) {
+  if (is_a($layout, 'Layout') && $layout->hasContexts(array('dashboard'))) {
```
Comment (herbdool) [inline core/modules/dashboard/dashboard.module:225]: @olafgrabienski I'm thinking we should have the same thing in the function just above this one. Can you make it:

```
function dashboard_preprocess_layout(&$variables) {
  /* @var Layout $layout */
  $layout = $variables['layout'];
  if (is_a($layout, 'Layout') && $layout->hasContexts(array('dashboard'))) {
```

Comment (olafgrabienski) [reply core/modules/dashboard/dashboard.module:225]: Sure, done!

**Comment by herbdool** [comment]
Failing test is unrelated.

**Comment by quicksketch** [comment]
Thanks @herbdool and @olafgrabienski for working this one out!

**Pull request merged**